Improve google calendar test coverage to 97%

allenporter opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->



## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Improve google calendar test coverage to 97% and fix one error on setup found from tests. There are additional error cases that appear to have inconsistencies or unintended bugs (e.g. missing keys not handled gracefully), but they can be fixed in follow ups.

Removed tests that were not exercising public entity APIs.
Removed tests that were mocking out large parts of the code.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [X] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] The code has been formatted using Black (`black --fast homeassistant tests`)
- [X] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

allenporter committed:
Improve google calendar test coverage to 97%

allenporter committed:
Remove commented out code.

allenporter committed:
Remove unnecessary (flaky) checks for token file persistence

allenporter committed:
Remove mock code assertions

allenporter committed:
Add debug logging to google calendar integration

allenporter committed:
Increase alarm time to polling code to reduce flakes

allenporter committed:
Setup every test in their own configuration directory

MartinHjelmare commented on tests/components/google/test_init.py:
```diff
@@ -1,67 +1,513 @@
+    mock_load_platform.assert_called_once()
+
+    # One calendar configured
+    assert len(hass.data[google.DATA_INDEX]) == 1
```
We shouldn't assert any details like hass.data. We can check that there's a state in the state machine instead.

allenporter replied:
Fixed, now looks at entities.

MartinHjelmare commented on tests/components/google/test_init.py:
```diff
@@ -1,67 +1,513 @@
+    hass: HomeAssistant, creds: OAuth2Credentials, token_filename: Path
+) -> None:
+    """Fixture to populate an existing token file."""
+    storage = Storage(token_filename)
```
Isn't it possible to patch out the client so we don't need to worry about writing files at all?

allenporter replied:
Removed. See comments below about additional code changes i had to make to keep this simple.

MartinHjelmare commented on tests/components/google/test_init.py:
```diff
@@ -1,67 +1,513 @@
+    yaml_devices_filename: Path,
+) -> None:
+    """Fixture that prepares the calendars.yaml file."""
+    with open(yaml_devices_filename, "w") as out:
```
We don't want to write real files I think. We could patch this out with some memory buffer if we really want to test the writing and reading from buffer. Otherwise, just patch the YAML loading/dumping and use a dict instead.

allenporter replied:
Yeah, the existing tests patch this out completely, so the idea is to actually increase test coverage.  So we'd prefer to patch the `open` call in the integration? for example https://github.com/home-assistant/core/blob/a4904bd9bc3bcd80e5f688e4cd017fb25594e397/homeassistant/components/google/__init__.py#L426

MartinHjelmare replied:
That could work. If it's tricky I don't think it's too bad if we patch one level higher.

Hopefully these files will go away when we can implement a trigger platform for calendars.

allenporter replied:
Yeah, i realize this is a lot of test code for something we plan to gut. It's just helpful for me to learn this code, and i never manually test anything. My impression is that the google calendar auth code performs pretty flaky, so just don't want to make it much worse.

I'll see, and if its bad we can give up on getting the extra test coverage, or i can move it into a separate file.

MartinHjelmare replied:
Sounds good.

allenporter replied:
It seems like it wasn't so bad.  However, i noticed the token files are read/written in multiple ways so i re-wrote the token reading code that checks for existing scopes. It now uses Storage.get just like the other code, and looks in the creds object.   I think I approved the code that added that token check, but wasn't as familiar with the code at the time so this is my fault.

allenporter committed:
Mock out filesystem calls to avoid disk dependencies

Update scope checking code to use Storage object rather than text file matching

allenporter committed:
Update tests to check entity states when integration is loaded

allenporter commented:
There is something else flaky here where the tests pass run separately but not together. (This was another advantage of wiping out the file state every run...) so I need to dig into this more. Hold off on further review.

allenporter committed:
Mock out google service in multiple locations

MartinHjelmare commented on homeassistant/components/google/__init__.py:
```diff
@@ -247,27 +249,25 @@ def setup(hass: HomeAssistant, config: ConfigType) -> bool:
-            _LOGGER.warning("Please re-authenticate with Google")
-            return False
-    return True
+    creds = Storage(hass.config.path(TOKEN_FILE)).get()
```
```suggestion
    creds = Storage(token_file).get()
```

MartinHjelmare commented on homeassistant/components/google/__init__.py:
```diff
@@ -247,27 +249,25 @@ def setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    creds = Storage(hass.config.path(TOKEN_FILE)).get()
+    if not creds or not creds.scopes:
+        return False
+    target_scope = config.get(CONF_CALENDAR_ACCESS).scope
```
There's a default for this config option in the config schema.
```suggestion
    target_scope = config[CONF_CALENDAR_ACCESS].scope
```

allenporter committed:
Update homeassistant/components/google/__init__.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

allenporter committed:
Update homeassistant/components/google/__init__.py

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

MartinHjelmare reviewed (approved):
Great!
